### PR TITLE
Abstract world

### PR DESCRIPTION
**Pull request by AbdelrhmanBassiouny**
Abstracting the world class from pybullet to have a flexible simulator backend for pycram

### CODE REVIEW

File: src/pycram/world.py [hunk 1]
```diff
@@ -0,0 +1,2070 @@
...
+
+    def get_positions_of_all_joints(self) -> Dict[str: float]:
+        """
+        Returns the positions of all joints of the object as a dictionary of joint names and joint positions.
+
+        :return: A dictionary with all joints positions'.
+        """
+        return self._current_joints_positions
+
+    def update_link_transforms(self, transform_time: Optional[rospy.Time] = None):
```
Comment (tomsch420) [inline src/pycram/world.py:1712]: DOC

File: src/pycram/world.py [hunk 2]
```diff
@@ -0,0 +1,2070 @@
...
+    :param file_name: The filename of the searched file.
+    :param data_directory: The directory in which to search for the file.
+    :return: The full path in the filesystem or None if there is no file with the filename in the directory
+    """
+    for file in os.listdir(data_directory):
+        if file == file_name:
+            return data_directory + f"/{file_name}"
+
+
+def _get_robot_name_from_urdf(urdf_string: str) -> str:
```
Comment (tomsch420) [inline src/pycram/world.py:1742]: remove it

File: src/pycram/world.py [hunk 3]
```diff
@@ -0,0 +1,2070 @@
...
+    def __init__(self,
+                 parent_link: Link,
+                 child_link: Link,
+                 bidirectional: Optional[bool] = False,
+                 parent_to_child_transform: Optional[Transform] = None,
+                 constraint_id: Optional[int] = None):
+        """
+        Creates an attachment between the parent object and the child object.
+        """
+        self.parent_link = parent_link
```
Comment (tomsch420) [inline src/pycram/world.py:1768]: typing

File: src/pycram/world.py [hunk 4]
```diff
@@ -0,0 +1,2070 @@
...
+        self.child_link = child_link
+        self.bidirectional = bidirectional
+        self._loose = False and not bidirectional
```
Comment (tomsch420) [inline src/pycram/world.py:1771]: loose should be be "not bidirectional"

File: src/pycram/world.py [hunk 5]
```diff
@@ -0,0 +1,2070 @@
...
+
+class Attachment:
+    def __init__(self,
+                 parent_link: Link,
+                 child_link: Link,
+                 bidirectional: Optional[bool] = False,
+                 parent_to_child_transform: Optional[Transform] = None,
+                 constraint_id: Optional[int] = None):
+        """
+        Creates an attachment between the parent object and the child object.
```
Comment (tomsch420) [inline src/pycram/world.py:1766]: no doc for the parameter

File: src/pycram/world.py [hunk 6]
```diff
@@ -0,0 +1,2070 @@
...
+
+        self.parent_to_child_transform = parent_to_child_transform
+        if self.parent_to_child_transform is None:
+            self.update_transform()
+
+        self.constraint_id = constraint_id
+        if self.constraint_id is None:
+            self.add_fixed_constraint()
+
+    def update_transform_and_constraint(self):
```
Comment (tomsch420) [inline src/pycram/world.py:1781]: DOC!!!!!111!

File: src/pycram/world.py [hunk 7]
```diff
@@ -0,0 +1,2070 @@
...
+        return self.parent_link.get_transform_to_link(self.child_link)
+
+    def remove_constraint_if_exists(self):
+        if self.child_link in self.parent_link.constraint_ids:
+            self.parent_link.remove_constraint_with_link(self.child_link)
+
+    def get_inverse(self):
+        attachment = Attachment(self.child_link, self.parent_link, self.bidirectional,
+                                constraint_id=self.constraint_id)
+        attachment.loose = False if self.loose else True
```
Comment (tomsch420) [inline src/pycram/world.py:1806]: change to "not self.loose"

File: src/pycram/world.py [hunk 8]
```diff
@@ -0,0 +1,2070 @@
...
+    rgb = " ".join(list(map(str, color.get_rgba())))
+    pathlib_obj = pathlib.Path(path)
+    path = str(pathlib_obj.resolve())
+    content = urdf_template.replace("~a", name).replace("~b", path).replace("~c", rgb)
+    with open(cach_dir + pathlib_obj.stem + ".urdf", "w", encoding="utf-8") as file:
+        file.write(content)
+    return cach_dir + pathlib_obj.stem + ".urdf"
+
+
+def _world_and_id(world: World) -> Tuple[World, int]:
```
Comment (tomsch420) [inline src/pycram/world.py:2061]: remove

File: src/pycram/world_dataclasses.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+
+@dataclass
+class Color:
+    """
+    Dataclass for storing color as an RGBA value.
+    """
+    R: float = 1
+    G: float = 1
+    B: float = 1
+    A: float = 1
```
Comment (tomsch420) [inline src/pycram/world_dataclasses.py:14]: doc A :tomato:

File: src/pycram/world_dataclasses.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    """
+    Dataclass for storing a constraint between two objects.
+    """
+    parent_obj_id: int
+    parent_link_name: str
+    child_obj_id: int
+    child_link_name: str
+    joint_type: JointType
+    joint_axis_in_child_link_frame: List[int]
+    joint_frame_position_wrt_parent_origin: List[float]
```
Comment (tomsch420) [inline src/pycram/world_dataclasses.py:45]: Use a Transform here

File: src/pycram/world_dataclasses.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    joint_frame_position_wrt_child_origin: List[float]
+    joint_frame_orientation_wrt_parent_origin: Optional[List[float]] = None
+    joint_frame_orientation_wrt_child_origin: Optional[List[float]] = None
+
+
+@dataclass
+class Point:
```
Comment (tomsch420) [inline src/pycram/world_dataclasses.py:52]: Use Point message, look at pose for example

File: src/pycram/world_dataclasses.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+@dataclass
+class MultiBody:
+    base_visual_shape_index: int
+    base_position: List[float]
+    base_orientation: List[float]
+    link_visual_shape_indices: List[int]
+    link_positions: List[List[float]]
+    link_orientations: List[List[float]]
+    link_masses: List[float]
+    link_inertial_frame_positions: List[List[float]]
```
Comment (tomsch420) [inline src/pycram/world_dataclasses.py:153]: Use Pose

File: src/pycram/world_reasoning.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+import numpy as np
+
+from .external_interfaces.ik import try_to_reach, try_to_reach_with_grasp
+from .pose import Pose, Transform
+from .robot_descriptions import robot_description
+from .world import Object, UseProspectionWorld
+from .world import World
+
+
+class ReasoningError(Exception):
```
Comment (tomsch420) [inline src/pycram/world_reasoning.py:13]: move to pan_failures

File: src/pycram/world_reasoning.py [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+            for point in con_points:
+                contact_links.append((prospection_obj1.get_link_by_id(point[3]),
+                                      prospection_obj2.get_link_by_id(point[4])))
+            return con_points != (), contact_links
+
+        else:
+            return con_points != ()
+
+
+def get_visible_objects(
```
Comment (tomsch420) [inline src/pycram/world_reasoning.py:78]: doc ...

**Review by tomsch420 — CHANGES_REQUESTED** [review]
Check the tutorials for intro and giskard if this is how it should be.

File: src/pycram/pose.py [hunk 15]
```diff
@@ -301,6 +304,10 @@ def __init__(self, translation: Optional[List[float]] = None, rotation: Optional
 
         self.frame = frame
 
+    @classmethod
```
Comment (tomsch420) [inline src/pycram/pose.py:307]: use rospy timestamp

File: src/pycram/process_modules/boxy_process_modules.py [hunk 16]
```diff
@@ -16,6 +17,15 @@
 from tf.transformations import euler_from_quaternion, quaternion_from_euler
 
 
+def _transform_to_torso(pose_and_rotation: Tuple[List[float], List[float]], robot: BulletWorldObject) -> Tuple[
```
Comment (tomsch420) [inline src/pycram/process_modules/boxy_process_modules.py:20]: delete this method

**Review by tomsch420 — CHANGES_REQUESTED** [review]
Furthermore check for code coverage.

**Review by tomsch420 — CHANGES_REQUESTED** [review]
Testing PR view in pycharm

File: src/pycram/world.py [hunk 17]
```diff
@@ -0,0 +1,2070 @@
...
+
+    def update_transform(self, transform_time: Optional[rospy.Time] = None):
+        self.local_transformer.update_transforms([self.transform], transform_time)
+
+    @property
+    def transform(self) -> Transform:
+        """
+        The transformation from the world frame to this link frame.
+        """
+        return self.pose.to_transform(self.tf_frame)
```
[the remark this answers is not in the document]
Comment (AbdelrhmanBassiouny) [reply src/pycram/world.py:1281]: it does not but there is a stamp in the header of the pose that can be used.

File: src/pycram/world.py [hunk 18]
```diff
@@ -856,11 +1010,15 @@ class UseProspectionWorld:
...
-        self.prev_world: World = None
+        self.prev_world: Optional[World] = None
+        # The previous world is saved to restore it after the with block is exited.
 
     def __enter__(self):
+        """
+        This method is called when entering the with block, it will set the current world to the prospection world
+        """
         if not World.current_world.is_prospection_world:
-            time.sleep(20 * World.simulation_time_step)
```
Comment (tomsch420) [inline src/pycram/world.py]: make this 20 a constant somewhere

File: src/pycram/world_constraints.py [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+
+
+class AbstractConstraint:
+    """
+    Represents an abstract constraint concept, this could be used to create joints for example or any kind of constraint
+    between two links in the world.
+    """
+
+    def __init__(self,
+                 parent_link: 'Link',
```
Comment (tomsch420) [inline src/pycram/world_constraints.py:15]: annotations

File: src/pycram/world_object.py [hunk 20]
```diff
@@ -0,0 +1,777 @@
+import logging
+import os
+
+import numpy as np
+import rospy
+from geometry_msgs.msg import Point, Quaternion
+from typing_extensions import Type, Optional, Dict, Tuple, List, Union
+
+from pycram.enums import ObjectType, JointType
```
Comment (tomsch420) [inline src/pycram/world_object.py:9]: absolute imports

**Comment by Tigul** [comment]
With [85eddde](https://github.com/cram2/pycram/commit/85eddde1cb76bf9fa1652f7a7ca493754e666742) this PR is merged. I don't know why GitHub does not recognize it but I will close this PR with this comment.

**Pull request closed without merging**